File: mpp-proiect-java-culbec/MotorcycleContest/Networking/src/main/java/NetworkUtils/ProtobufServer.java
package NetworkUtils;

import CommonUtils.IService;
import Workers.RPCReflectionWorker;

import java.net.Socket;

public class ProtobufServer extends AbstractConcurrentServer {
    private final IService server;

    public ProtobufServer(String hostname, int port, IService server) {
        super(hostname, port);
        this.server = server;
    }

    @Override
    protected Thread createWorker(Socket client) throws ServerException {
        // Creating a worker specific for this type of server.
        logger.traceEntry("Creating a new worker for the client...");

        try {
            RPCReflectionWorker worker = new RPCReflectionWorker(server, client);
            logger.traceExit("Worker created.");
            return new Thread(worker);
        } catch (Exception e) {
            logger.error("Couldn't create the worker: {}", e.getMessage());
            throw new ServerException(e.getMessage());
        }
    }
}

File: mpp-proiect-csharp-culbec/Network/NetworkUtils/ProtobufServer.cs
using System.Net.Sockets;
using Common.CommonUtils;
using Common.Exceptions;
using log4net;
using Network.Workers;

namespace Network.NetworkUtils;

public class ProtobufServer : AbstractConcurrentServer
{
    private static readonly ILog Log = LogManager.GetLogger(typeof(ProtobufServer));
    private readonly IService _server;
    public ProtobufServer(string host, int port, IService server) : base(host, port)
    {
        _server = server;
    }

    protected override Thread CreateWorker(TcpClient client)
    {
        Log.Info("Creating a new worker...");
        ProtobufWorker worker;

        try
        {
            worker = new ProtobufWorker(_server, client);
        }
        catch (Exception e)
        {
            Log.ErrorFormat("An error occurred: {0}", e.Message);
            throw new AppException(e.Message);
        }

        Log.Info("Worker created.");
        return new Thread(worker.Run);
    }
}